src/pEpEngine.c

```diff
@@ -4176,11 +4177,277 @@ PEP_STATUS set_all_userids_to_own(PEP_SESSION session, identity_list* id_list) {
+    PEP_STATUS status = own_identity_by_address(session, account, &found_identity);
+
+    if (status != PEP_STATUS_OK) {
+        return status;
+    }
+
+    // Ensure that session doesn't contain a previous passphrase.
+    status = config_passphrase(session, NULL);
+    if (status != PEP_STATUS_OK) {
+        free_identity(found_identity);
+        return status;
+    }
+
+    status = sign_something(session, found_identity);
+    free_identity(found_identity);
+    if (status == PEP_STATUS_OK) {
+        *has_passphrase = false;
+        return PEP_STATUS_OK;
+    } else if (status == PEP_PASSPHRASE_REQUIRED) {
+        *has_passphrase = true;
+        return PEP_STATUS_OK;
+    }
+
+    return PEP_CANNOT_FIND_IDENTITY;
+}
+
+DYNAMIC_API PEP_STATUS
+unlock_keys_with_passphrase(PEP_SESSION session,
+                            const stringpair_list_t *accounts_with_passphrases,
+                            stringlist_t **error_accounts)
+{
+    PEP_REQUIRE(accounts_with_passphrases);
+    PEP_REQUIRE(error_accounts);
+    *error_accounts = NULL;
+
+    PEP_STATUS resulting_status = PEP_CANNOT_FIND_IDENTITY;
+
+    for (stringpair_list_t *account_passphrase_pair = accounts_with_passphrases;
+         account_passphrase_pair;
+         account_passphrase_pair = account_passphrase_pair->next) {
+        const stringpair_t *current = account_passphrase_pair->value;
+
+        if (!current) {
+            free_stringlist(*error_accounts);
+            return PEP_ILLEGAL_VALUE;
+        }
+
+        if (!current->key) {
+            free_stringlist(*error_accounts);
+            return PEP_ILLEGAL_VALUE;
+        }
+
+        if (!current->value) {
+            free_stringlist(*error_accounts);
+            return PEP_ILLEGAL_VALUE;
+        }
+
+        pEp_identity *found_identity = NULL;
+        PEP_STATUS find_status = own_identity_by_address(session, current->key, &found_identity);
+        if (find_status != PEP_STATUS_OK) {
+            if (*error_accounts) {
+                free_stringlist(*error_accounts);
+            }
+            *error_accounts = new_stringlist(current->key);
+            return PEP_CANNOT_FIND_IDENTITY;
+        }
+
+        PEP_STATUS config_passphrase_status = PEP_STATUS_OK;
+        if (EMPTYSTR(current->value)) {
+            // An empty or NULL passphrase means to not use any passphrase.
+            config_passphrase_status = config_passphrase(session, NULL);
+        } else {
+            config_passphrase_status = config_passphrase(session, current->value);
+        }
+        if (config_passphrase_status != PEP_STATUS_OK) {
+            free_stringlist(*error_accounts);
+            free_identity(found_identity);
+            return config_passphrase_status;
+        }
+
+        PEP_STATUS sign_status = sign_something(session, found_identity);
+        free_identity(found_identity);
+
+        // Remove any passphrase from the session.
+        config_passphrase_status = config_passphrase(session, NULL);
```

Comment:
why do you remove the passphrase from the session again? shouldn't it be kept *in the session* if it was successful, so any follow attempt later to reset a key or generate a new key will result in using the previously entered passphrase?

Reply:
This is purely internal, trying to keep it tidy.

The actual caching happens in the wrapper, and the passphrase is set _preemptively_, e.g. like this:

```
namespace pEp {
    template<typename... A>
    PEP_STATUS PassphraseCache::api(PEP_STATUS f(PEP_SESSION, A...), PEP_SESSION session, A... a)
    {
        PEP_STATUS status{ PEP_STATUS_OK };

        for_each_passphrase([&](const std::string& passphrase) {
            status = ::config_passphrase(session, passphrase.c_str());
            if (status) {
                return true;
            }

            status = f(session, a...);
            return status != PEP_PASSPHRASE_REQUIRED && status != PEP_WRONG_PASSPHRASE;
        });

        return status;
    }
}
```